Mathlib/Algebra/Category/ModuleCat/Monoidal/Closed.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Scott Morrison, Jakob von Raumer
+
+! This file was ported from Lean 3 source module algebra.category.Module.monoidal.closed
+! leanprover-community/mathlib commit 74403a3b2551b0970855e14ef5e8fd0d6af1bfc2
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Closed.Monoidal
+import Mathlib.Algebra.Category.ModuleCat.Monoidal.Symmetric
+
+/-!
+# The monoidal closed structure on `Module R`.
+-/
+
+
+universe v w x u
+
+open CategoryTheory Opposite
+
+namespace ModuleCat
+
+variable {R : Type u} [CommRing R]
+
+-- porting note: removed @[simps] as the simpNF linter complains
+/-- Auxiliary definition for the `MonoidalClosed` instance on `Module R`.
+(This is only a separate definition in order to speed up typechecking. )
+-/
+def monoidalClosedHomEquiv (M N P : ModuleCat.{u} R) :
+    ((MonoidalCategory.tensorLeft M).obj N ⟶ P) ≃
+      (N ⟶ ((linearCoyoneda R (ModuleCat R)).obj (op M)).obj P) where
+  toFun f := LinearMap.compr₂ (TensorProduct.mk R N M) ((β_ N M).hom ≫ f)
+  invFun f := (β_ M N).hom ≫ TensorProduct.lift f
+  left_inv f := by
+    apply TensorProduct.ext'
+    intro m n
+    rw [coe_comp, Function.comp_apply, MonoidalCategory.braiding_hom_apply,
+      TensorProduct.lift.tmul, LinearMap.compr₂_apply,
+      TensorProduct.mk_apply, coe_comp, Function.comp_apply, MonoidalCategory.braiding_hom_apply]
+  right_inv f := rfl
+set_option linter.uppercaseLean3 false in
+#align Module.monoidal_closed_hom_equiv ModuleCat.monoidalClosedHomEquiv
+
+instance : MonoidalClosed (ModuleCat.{u} R) where
+  closed M :=
+    { isAdj :=
+        { right := (linearCoyoneda R (ModuleCat.{u} R)).obj (op M)
+          adj := Adjunction.mkOfHomEquiv
+            { homEquiv := fun N P => monoidalClosedHomEquiv M N P
+              -- porting note: this proof was automatic in mathlib3
+              homEquiv_naturality_left_symm := by
+                intros
+                apply TensorProduct.ext'
+                intro m n
+                rfl } } }
+
+theorem ihom_map_apply {M N P : ModuleCat.{u} R} (f : N ⟶ P) (g : ModuleCat.of R (M ⟶ N)) :
+    (ihom M).map f g = g ≫ f :=
+  rfl
+set_option linter.uppercaseLean3 false in
+#align Module.ihom_map_apply ModuleCat.ihom_map_apply
+
+-- porting note: `CoeFun` was replaced by `FunLike`
+-- I can't seem to express the function coercion here without writing `@FunLike.coe`.
```

Comment:
Hmm, this is awkward, but I'm not finding anything better. Shall we merge and hope for the best?